.github/workflows/ci.yml

```diff
@@ -0,0 +1,213 @@
+      fail-fast: false
+    steps:
+      - run: echo "Running on $RUNNER_OS $RUNNER_ARCH"
+      - uses: actions/checkout@v4
+        with:
+          submodules: recursive
+      - uses: ./.github/actions/configure-environment
+      - if: github.event.inputs.ref != ''
+        uses: actions/checkout@v4
+        with:
+          submodules: recursive
+          ref: ${{ github.event.inputs.ref }}
+      - if: runner.os == 'macOS'
+        run: cd rust && cargo fetch
+      - name: Build project
+        run: make
+      - name: Build project without CGO
+        run: env CGO_ENABLED=0 go build .
+      - if: runner.os == 'Linux'
+        uses: actions/cache/restore@v3
+        with:
+          key: v28-proof-params-${{ runner.os }}-${{ runner.arch }}
+          path: |
+            ./filecoin-proof-parameters
+      - if: runner.os == 'Linux'
+        name: Obtain Filecoin parameters
+        run: |
+          DIR=$(pwd)
+          cd $(mktemp -d)
+          go install github.com/filecoin-project/go-paramfetch/paramfetch@latest
+          $GOPATH/bin/paramfetch 2048 "${DIR}/parameters.json" "${DIR}/srs-inner-product.json"
+      - if: runner.os == 'Linux' && ((github.event == 'push' && !startsWith(github.ref, 'refs/tags/')) || github.event.inputs.save == 'true')
+        uses: actions/cache/save@v3
+        with:
+          key: v28-proof-params-${{ runner.os }}-${{ runner.arch }}
+          path: |
+            ./filecoin-proof-parameters
+      - if: runner.os == 'Linux'
+        run: cd rust && rustup target add wasm32-unknown-unknown
+      - if: github.event.inputs.run-leak-detector == 'true'
+        run: make cgo-leakdetect
+      - if: runner.os == 'Linux'
+        run: cd rust && FIL_PROOFS_PARAMETER_CACHE="${GITHUB_WORKSPACE}/filecoin-proof-parameters/" RUST_LOG=info cargo test --all --release -- --test-threads 1 && cd ..
+      - if: runner.os == 'Linux'
+        run: GOEXPERIMENT=cgocheck2 RUST_LOG=info go test -p 1 -timeout 60m
+      - if: runner.os == 'macOS'
+        name: Build project and tests, but don't actually run the tests (used to verify that build/link works with Darwin)
+        run: GOEXPERIMENT=cgocheck2 RUST_LOG=info go test -run=^$
+  supraseal:
+    name: Build with CUDA supraseal
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: recursive
+      - uses: ./.github/actions/configure-environment
+      - if: github.event.inputs.ref != ''
+        uses: actions/checkout@v4
+        with:
+          submodules: recursive
+          ref: ${{ github.event.inputs.ref }}
+      - name: Build project with `FFI_USE_CUDA_SUPRASEAL=1`
+        run: FFI_BUILD_FROM_SOURCE=1 FFI_USE_CUDA_SUPRASEAL=1 make
+  publish:
+    needs: [check, cgo-bindings, supraseal]
+    if: (github.event_name == 'push' && startsWith(github.ref, 'refs/tags/')) || github.event.inputs.publish == 'true'
```

Comment:
Is there no nicer way to specify "only run on `v.*` tag?"? Both this and its inverse above are fine, the intent is hidden under layers of abstraction which add to the general incomprehensibility of GHA config language.

Reply:
I'm afraid that's the most straighforward way of making this check at the moment unless we wanted to split the workflow into multiple files.